packages/loopring_v3/contracts/impl/agents/FastWithdrawalAgent.sol

```diff
@@ -0,0 +1,203 @@
+    }
+
+    function executeFastWithdrawals(FastWithdrawal[] memory fastWithdrawals)
+        public
+        payable
+        nonReentrant
+    {
+        // Do all fast withdrawals
+        for (uint i = 0; i < fastWithdrawals.length; i++) {
+            executeFastWithdrawal(fastWithdrawals[i]);
+        }
+        // Return any ETH left into this contract
+        // (can happen when more ETH is sent than needed for the fast withdrawals)
+        msg.sender.sendETHAndVerify(address(this).balance, gasleft());
+    }
+
+    // -- Internal --
+
+    function executeFastWithdrawal(FastWithdrawal memory fastWithdrawal)
+        internal
+    {
+        // Currently here because the compiler hangs when doing this in the constructor
+        if (DOMAIN_SEPARATOR == bytes32(0)) {
+            DOMAIN_SEPARATOR = EIP712.hash(EIP712.Domain("FastWithdrawalAgent", "1.0", address(this)));
+        }
+
+        // Compute the hash
+        bytes32 hash = EIP712.hashPacked(
+            DOMAIN_SEPARATOR,
+            keccak256(
+                abi.encodePacked(
+                    FASTWITHDRAWAL_TYPEHASH,
+                    fastWithdrawal.exchange,
+                    fastWithdrawal.from,
+                    fastWithdrawal.to,
+                    fastWithdrawal.token,
+                    fastWithdrawal.amount,
+                    fastWithdrawal.feeToken,
+                    fastWithdrawal.fee,
+                    fastWithdrawal.onchainFeePayment,
+                    fastWithdrawal.validUntil
+                )
+            )
+        );
+
+        // Make sure the tx isn't used multiple times
+        require(txHash[hash] == false, "TX_REPLAY");
+        txHash[hash] = true;
+
+        // Check the signature
+        require(hash.verifySignature(fastWithdrawal.from, fastWithdrawal.signature), "INVALID_SIGNATURE");
+
+        // Check the time limit
+        require(now <= fastWithdrawal.validUntil, "TX_EXPIRED");
+
+        // Transfer the tokens to the requested address
+        // using funds from `msg.sender`.
+        transfer(
+            fastWithdrawal.to,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // If the fee is paid offchain in the same token as the main transfer
+        // just add it to the transfer amount
+        if (!fastWithdrawal.onchainFeePayment && fastWithdrawal.token == fastWithdrawal.feeToken) {
+            fastWithdrawal.amount = fastWithdrawal.amount.add(fastWithdrawal.fee);
+            fastWithdrawal.fee = 0;
+        }
+
+        IExchangeV3 exchange = IExchangeV3(fastWithdrawal.exchange);
+
+        // Approve the offchain transfer from the withdrawing account back to the mediator (msg.sender)
+        exchange.approveConditionalTransfer(
+            fastWithdrawal.from,
+            msg.sender,
+            fastWithdrawal.token,
+            fastWithdrawal.amount
+        );
+
+        // Fee payment to the mediator (msg.sender)
```

Comment:
The fee is sent to get mediator because it's really him that provides the main service. But the assumption here is that the mediator is the same entity as the operator, or they at least have some kind of agreement (either offchain or onchain in the mediator/liquidity contract).

Perhaps it could be useful to be able to pay a fee to the a different address, I'm not sure it would really get used though.

Reply:
I agree that mediators will have to trust the operator in that the user has enough balance offchain to pay back. But other than that, I still think it's good for the mediators to be passive by not singing any transactions. Then you and I will easily become mediators by just providing funds based on the trust, there is no investment in building a small service or provide private keys to the operator.

Reply:
> It seems the `approveConditionalTransfer` for the operator to get the fee is not called.

What I meant is in the code below (inside ExchangeBlocks.sol):
```

    function consumeConditionalTransfer(
        ExchangeData.State storage S,
        uint24 operatorAccountID,
        uint transferData
        )
        private
    {
        // Extract the transfer data
        uint24 fromAccountID = uint24((transferData >> 80) & 0xFFFFFF);
        uint24 toAccountID = uint24((transferData >> 56) & 0xFFFFFF);
        uint16 tokenID = uint16((transferData >> 48) & 0xFF);
        uint amount = ((transferData >> 24) & 0xFFFFFF).decodeFloat(24);
        uint16 feeTokenID = uint16((transferData >> 16) & 0xFF);
        uint feeAmount = ((transferData >> 0) & 0xFFFF).decodeFloat(16);

        // Update onchain approvals
        if (amount > 0) {
            S.approvedTransferAmounts[fromAccountID][toAccountID][tokenID] =
                S.approvedTransferAmounts[fromAccountID][toAccountID][tokenID].sub(amount);
            emit ConditionalTransferConsumed(fromAccountID, toAccountID, tokenID, amount);
        }
        if (feeAmount > 0) {
            S.approvedTransferAmounts[fromAccountID][operatorAccountID][feeTokenID] =
                S.approvedTransferAmounts[fromAccountID][operatorAccountID][feeTokenID].sub(feeAmount);
            emit ConditionalTransferConsumed(fromAccountID, operatorAccountID, feeTokenID, feeAmount);
        }
    }
```

from ->operator conditional transfer approval is consumed, but I didn't find the corresponding approve invitation here in this file.

**The assumption that "mediator" is "operator" is too implicit in the code.** We should make it clear that it must be the operator who call the function and the mediator risks their assets based on trusts.

Reply:
> I agree that mediators will have to trust the operator in that the user has enough balance offchain to pay back. But other than that, I still think it's good for the mediators to be passive by not singing any transactions. Then you and I will easily become mediators by just providing funds based on the trust, there is no investment in building a small service or provide private keys to the operator.

In all cases the mediator will need to have some whitelist of accounts which are allowed to spend funds. The way I see this working is that in all cases the funds will be stored in a contract because of the additional security properties. So the mediator will simply whitelist the operator in a way that he can only do the fast withdrawals (or something else) via the contract. So the mediator is completely passive, it will just need to whitelist the operator to spend its funds in some way.

Reply:
I think the whitelist idea will work. Anyway, I think if we introduce the "mediator" role, it should be distinct from the "operator" role, and the code need to clearly reflect their relationship.